x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

```diff
@@ -234,26 +246,43 @@ private static QueryBuilder getApplicationNameQuery(Collection<String> applicati
 
         TermsQueryBuilder termsQuery = rawNames.isEmpty() ? null : QueryBuilders.termsQuery(APPLICATION.getPreferredName(), rawNames);
         if (wildcardNames.isEmpty()) {
-            return termsQuery;
-        }
-        final BoolQueryBuilder boolQuery = QueryBuilders.boolQuery();
-        if (termsQuery != null) {
-            boolQuery.should(termsQuery);
+            return new Tuple<>(termsQuery, null);
         }
-        for (String wildcard : wildcardNames) {
-            final String prefix = wildcard.substring(0, wildcard.length() - 1);
-            boolQuery.should(QueryBuilders.prefixQuery(APPLICATION.getPreferredName(), prefix));
+
+        if (allowExpensiveQueries) {
+            final BoolQueryBuilder boolQuery = QueryBuilders.boolQuery();
+            if (termsQuery != null) {
+                boolQuery.should(termsQuery);
+            }
+            for (String wildcard : wildcardNames) {
+                final String prefix = wildcard.substring(0, wildcard.length() - 1);
+                boolQuery.should(QueryBuilders.prefixQuery(APPLICATION.getPreferredName(), prefix));
+            }
+            boolQuery.minimumShouldMatch(1);
+            return new Tuple<>(boolQuery, null);
+        } else {
+            logger.trace("expensive queries are not allowed, switching to filtering application names in memory");
+            return new Tuple<>(QueryBuilders.existsQuery(APPLICATION.getPreferredName()), StringMatcher.of(applications));
```

Comment:
In the current implementation , I am pretty sure this is equivalent to a match all query since (for this doc type) that field will always exist. I think it would be more efficient to just do match all here and rely on the filter to provide all application privileges. 

However,  the in-memory post filtering (regardless of exists or match all query) is not ideal. Instead of post filtering it should be possible to configure the mappings to keep the keyword , but also add the index_prefix [mapping](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-prefixes.html) or use [edge ngrams](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-edgengram-tokenizer.html). I think if we introduce this now we can handle the limitations passively. 

For example, we could introduce index_prefix mapping and (when there is a trailing *)  if the prefix is less than 20 we always query against the index_prefix else if greater than 20 (max size for index_prefix) we fallback to the prefix query...unless expensive queries are disabled , in that case (greater than 20 with trailing * and no expensive queries) we fail with a message stating they need to make the application name shorter or enable expensive queries - that is passive since that failure condition (this bug) already exists today.  I think it would be exceptionally rare to hit the conditions to trigger the bug... so it would only be a partial fix to the bug that only fixes the bug when the application name is less than 20 characters. I think that is good enough, and the change would mostly remove the need to do expensive queries and introduce a meaningful and actionable error message.

Reply:
> In the current implementation , I am pretty sure this is equivalent to a match all query since (for this doc type) that field will always exist. I think it would be more efficient to just do match all here and rely on the filter to provide all application privileges.

Yes, a match-all query should do the job. This was a copy/paste from existing code without too much thinking. Now if we take one step further, we can just return null here since no further limiting query is needed since the caller already has the filter query (by doc type).

> we could introduce index_prefix mapping and (when there is a trailing *) if the prefix is less than 20 we always query against the index_prefix else if greater than 20 

This could be a good solution for a new index. But it does not help existing deployments with existing data. So I think it is not really viable here.

Reply:
>  But it does not help existing deployments with existing data. So I think it is not really viable here.

Hmm... this highlights a bigger problem that we don't have a strategy for mappings updates (even passive additions) outside of a major versions bumps. Technically what i describe is possible (with inplace mapping updates + fallback queries or a new index), but without that strategy/code support I agree it is not really viable. I think we should work towards defining that strategy but that is well beyond the scope of this PR. 

> we can just return null here since no further limiting query is needed

++

Reply:
Thanks for the comment. I wrote my previous comment in a hurry. So please let me explain myself a bit more.

We have limited support for mappings update. It is possible to [add new fields](https://github.com/elastic/elasticsearch/pull/82700/files#diff-debc2ea327ce1c4830bf0b06da8c2175e1ea03fa42a38e765f6f456f6e2baa00) to the mappings. These new fields will have `null` values for existing documents, i.e. there is no automatic process to populate the new fields for existing documents. We also cannot update existing fields. 

For the case here, we cannot update existing `application` field because (1) it is just not updatable (2) it is a `keyword` field while `_index_prefixes` requires `text`. So we will have to add a new field which would incur rather big overhead including both indexing and querying side changes. Also because this new field will not be populated for existing documents, we need to handle it in code to conditionally query the new field, if no result (this check could be tricky on its own), then fallback. It is not impossible. But the complexity does not feel right for this particular issue. In addition, being able to use expensive queries internally is a recurring ask. So a simpler solution seems to be more suitable because once internal expensive query usage is no longer an issue we can easily remove the workaround.

> I think we should work towards defining that strategy

I agree. It would be great if we could update existing field's mapping. Or if not, at least being able to populate fields (new or existing) for existing documents. Lack of this process has been causing frictions for a while. A recent occurrence is about the API key's `type` field which is an existing field but was never used till the recent cross-cluster API key work. It is now a bit of pain to support query this field because we want to fetch API keys with both `null` and `rest` types when user asks for `rest`. You can refer to ES-5990 if interested in details.